### DISCUSSION TITLE
Why don't use type embedding to load custom type collection records (when using pocketbase as framework)

### DISCUSSION THREAD

**Discussion by alexisvisco** [pocketbase/pocketbase#3064] [Q&A]
Hello dear mainainers ! 

### Question
Pocketbase can be used as a framework so I am wondering why there is no support for using the golang types in the dao api ? It would allow to benefits from the internal record API and other related hooks.

Using string to get,set fields seems really annoying and error prone. 

### What I would like to have

Can we imagine something like : 

```
type CustomRecord struct {
	*models.Record
	DiscordUsername string `db:"discord_username" json:"discord_username"`
}
var record CustomRecord
record, err := e.App.Dao().FindCustomRecordById("users", "80larfsmyukud28", &record)
if err != nil {
	return fmt.Errorf("unable to find article: %w", err)
}
record.DiscordUsername = "aaa"
e.App.Dao().Save(record)
```

We could simplify this when generics will allors type parameter on methods to have : 
`rec, err := e.App.Dao().FindRecordById[CustomRecord]("users", "80larfsmyukud28")`

### Current  workaround (not working)
Currently I was imagining a workaround with that solution : 
```
type CustomRecord struct {
	*models.Record
	DiscordUsername string `db:"discord_username" json:"discord_username"`
}
var record CustomRecord
collection, err := dao.FindCollectionByNameOrId("users")
if err != nil {
	return err
}
query := dao.RecordQuery(collection).AndWhere(dbx.HashExp{collection.Name + ".id": "80larfsmyukud28"})
if err := query.Limit(1).One(&record); err != nil {
	return err
}

// record.DiscordUsername is loaded as expected ! 

record.DiscordUsername = "test"

// now try to update it
// then print with the getter method 
fmt.Println(record.Get("discord_username")) // it panics here 


err = dao.Save(&record)
if err != nil {
	return err
}
```

<img width="853" alt="image" src="https://github.com/pocketbase/pocketbase/assets/17145294/d0276daf-e948-4c95-af9a-c3ca7a04255d">

It seems like data are not loaded into the record object, but the data are loaded from the tags structs. 

Do you have any insight why 🤔 ?

**Comment by alexisvisco** [comment]
This might be a lightest alternatives and cheaper (in term of code factorisation) to https://github.com/pocketbase/pocketbase/discussions/2697#discussioncomment-6154067

**Comment by alexisvisco** [comment]
> This might be a lightest alternatives and cheaper (in term of code factorisation) to [#2697 (reply in thread)](https://github.com/pocketbase/pocketbase/discussions/2697#discussioncomment-6154067)

But if it's not the case I can maybe implement a start of a codegen for what it's described in the discussion

**Comment by ganigeorgiev** [comment]
> It seems like data are not loaded into the record object, but the data are loaded from the tags structs.

Structs embedding is not inheritance. It panics because you don't have a initialized `models.Record` (it will be `nil` in your case and that is true also for its pointer fields) and the record model will have to be restructured if want to be used safely this way.

To answer your question why we don't use generics and embedding for the records - because I didn't found an easy way to do it. Suggestions for alternative approaches are welcomed but I don't see at the moment how the above will work without breaking everything and requiring users to create types for each collection.

Some of the existing implementation I'd do differently if we rewound but we carry the weight of being open source and I don't like introducing breaking changes even for pre-v1 unless the benefits are clear and really necessary.
